Enhancement : add schematics for ui-jar

PastaBolo opened the pull request:
Hello Kristoffer,

Ui-jar is such an awesome tool 😄 

I was learning Angular schematics and i thought that it would be useful to integrate them within ui-jar to avoid the manual set up with the ```ng-add ui-jar``` command.

I also added another schematics to support different ui-jar projects for multiple projects and libraries created within an angular workspace.

You can find more informations inside the ```./schematics/README.md``` file.

This is the first draft and it will certainly need improvments, but these schematics are some first ideas to improve ui-jar 😃 

What do you think ?

nording commented:
Sounds great, I will check it out later this week!

nording commented:
@PastaBolo when I trying to run this I get "Invalid rule result: Function()." - not sure what I'm doing wrong but maybe you can help me?

PastaBolo commented:
@nording 
That's

PastaBolo commented:
In fact, it doesn't work with Angular 6 😢 

I'd like to answer : run this command : `npm i -g @angular/cli@latest` 😆 
But I fixed it for Angular 6 compatibilty

Typically, the schematic do these steps :

- creates automatically a new project named ui-jar by calling the angular schematics `ng g application ui-jar`
- adds ui-jar dependency (you don't have to add it manually : `npm install ui-jar` is not needed)
- setups the project (updates the files so ui-jar works)

The last step is causing the issue because it replaces the 3 files that you listed with the ones to make ui-jar working.

---

### Correction (should now work with Angular 6)

I fixed my method which replace the files listed in the error you got.

Don't know why it works with Angular 7 but not with Angular 6. Maybe because schematics are still too young and sensible to big changes.

---

I will create some issues in my repo for future developments to improve the use of ui-jar-schematics

I let you try 😉

nording commented:
Nice work! :smile: 
For some reason I got some weird version number in my package.json, I got:

```
...
"ui-jar": "*",
...
```
Not sure why?
After I fixed that it worked like expected!

It would be nice if we could get this into next release - I really like how simple the installation is!

PastaBolo commented:
Thanks ! 😄 

For the version, I put '*' so npm would download the lastest version of ui-jar. Of course, this is ugly in package.json. I will update this to display the latest version by making first a request to https://registry.npmjs.org/ui-jar/ and getting the version with they `latest` key in the result.
I add an issue in my repo, I'll work on this later (this weekend I think) 😉 

But, if we integrate the ui-jar-schematics package directly in ui-jar, getting the version first won't be needed anymore because it will be possible to add ui-jar directly with the schematics `ng add ui-jar` : this will install ui-jar with the last version and run the schematics to set up automatically the ui-jar project

nording commented:
Ah okey so that's why the "*" is in package.json :) it would be cool to integrate this schemantics in ui-jar!

PastaBolo commented:
I made some updates :

- the version is now correctly displayed in the package.json file (a request is made to the npm api to get the latest version)
- ng add command is now supported

Now, instead of running :
```
npm i -D ui-jar schematics
ng g ui-jar-schematics:generate-uijar-project
```
you can simply run :
```
ng add ui-jar-schematics
```
This command will do all the steps from downloading ui-jar to setting up an initial ui-jar project 😉 

Next steps :
- little refactoring (https://github.com/PastaBolo/ui-jar-schematics/issues/2)
- creation of an external package to handle the utility functions (in utility folder)
- adjust the collection of schematics so they fit well with ui-jar when they will be integrated within it

We will make it ! 💪

picheli20 commented on schematics/README.md:
```diff
@@ -0,0 +1,49 @@
+# Ui-jar Schematics
```
nitpik: Would be good add some reference of this on the main README.md, so it don't get lost :D

picheli20 commented:
Looks awesome! Just one last thing, would be a feature, but we can add some questions regarding the customization for `UIJarConfiguration` and already add it :)

PastaBolo commented:
I'm sorry, really busy those days. 
That's a great idea, I'll work on this 😉
